[PodSecurity] Expand unit test coverage and fix error cases

tallclair opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

1. Add unit test for `ValidatePod`, exercising all possible response paths.
2. Reuse the test fixtures to refactor (and expand coverage) of `ValidatePodController` tests
3. Tests revealed that validation errors on pod controllers reject the request, even though controller validation is never enforcing. Errors validating controller resources now always return an allowed response.
4. Always set the `error` audit annotation when an error prevented normal policy evaluation.

#### Special notes for your reviewer:

This PR has overlap with several in-flight PRs, so conflicts will need to be resolved depending on which merges first.

#### Does this PR introduce a user-facing change?
```release-note
(PodSecurity admission) errors validating workload resources (deployment, replicaset, etc.) no longer block admission.
```

/sig auth
/assign @liggitt 
/milestone v1.23

tallclair committed:
[PodSecurity] Add ValidatePod unit test

tallclair committed:
[PodSecurity] Errors validating PodControllers are non blocking

tallclair committed:
[PodSecurity] Include error audit annotation on all non-forbidden errors

liggitt commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -389,12 +393,20 @@ func (a *Admission) ValidatePodController(ctx context.Context, attrs api.Attribu
-		return badRequestResponse("failed to decode object")
+		response := allowedResponse()
+		response.AuditAnnotations = map[string]string{
+			"error": "failed to decode object",
```
should we include the actual error in the audit log?

liggitt commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -389,12 +393,20 @@ func (a *Admission) ValidatePodController(ctx context.Context, attrs api.Attribu
-		return badRequestResponse("failed to extract pod template")
+		response := allowedResponse()
+		response.AuditAnnotations = map[string]string{
+			"error": "failed to extract pod template",
```
should we include the actual error in the audit log?

liggitt commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -379,7 +379,11 @@ func (a *Admission) ValidatePodController(ctx context.Context, attrs api.Attribu
-		return internalErrorResponse(fmt.Sprintf("failed to lookup namespace %s", attrs.GetNamespace()))
+		response := allowedResponse()
+		response.AuditAnnotations = map[string]string{
+			"error": fmt.Sprintf("failed to lookup namespace %s", attrs.GetNamespace()),
```
should we include the actual error in the audit log?

liggitt commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -603,18 +615,23 @@ func invalidResponse(attrs api.Attributes, fieldErrors field.ErrorList) *admissi
 // Relevant mutable pod fields as of 1.21 are image and seccomp annotations:
 // * https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/apis/core/validation/validation.go#L3947-L3949
 func isSignificantPodUpdate(pod, oldPod *corev1.Pod) bool {
+	// TODO: invert this logic to allow specific update types.
```
I'm not sure I understand the TODO

tallclair replied:
Currently we're enumerating the cases that call for re-evaluating the policy. I'd rather enumerate the cases that we can safely ignore, incase we ever relax the mutation restrictions in the future (or just missed something...). It makes the logic more complicated here, but means a safer failure scenario.

liggitt replied:
ah, ok

liggitt commented on staging/src/k8s.io/pod-security-admission/api/helpers.go:
```diff
@@ -156,6 +156,9 @@ func PolicyToEvaluate(labels map[string]string, defaults Policy) (Policy, field.
 
 		p = defaults
 	)
+	if labels == nil {
```
nit: expand to `if len(labels) == 0`?

liggitt reviewed (commented):
a few nits, lgtm overall

tallclair commented on staging/src/k8s.io/pod-security-admission/admission/admission_test.go:
```diff
@@ -191,12 +196,22 @@ func (t *testEvaluator) EvaluatePod(lv api.LevelVersion, meta *metav1.ObjectMeta
-type testNamespaceGetter struct {
-	ns *corev1.Namespace
-}
+// type testNamespaceGetter struct {
```
oops, done.

tallclair reviewed (commented):
Feedback addressed. Also refactored error response helpers a bit (including moving to `admission/response.go`)

tallclair commented:
This will need to be rebased once https://github.com/kubernetes/kubernetes/pull/105898 merges. Feel free to hold off on further review until I've had a chance to resolve that.

tallclair committed:
[PodSecurity] Add metrics test coverage

tallclair commented:
Rebased, and added additional coverage for error & exemption metrics. Also added a shared response for the privileged short-circuit case that includes the `enforce-policy` audit annotation, and refactored the shared responses a bit.

liggitt commented:
/retest
/lgtm

liggitt commented:
sigh... one last rebase needed

tallclair commented:
rebased

liggitt commented:
/lgtm
